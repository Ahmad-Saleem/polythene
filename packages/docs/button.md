# Button

Displays a text button with a [ripple](../polythene-ripple) effect.

See also: [Raised Button](../polythene-raised-button), [Icon Button](../polythene-icon-button) and [Floating Action Button](../polythene-fab)



## Usage

~~~javascript
import m from "mithril";
import button from "polythene-mithril-button";

const myButton = m(button, {
  label: "Button"
});
~~~

Add a URL:

~~~javascript
const myButton = m(button, {
  label: "Button",
  url: {
    href: "/index",
    oncreate: m.route.link
  }
});
~~~

Add an onclick event:

~~~javascript
const myButton = m(button, {
  label: "Button",
  events: {
    onclick: () => console.log("click")
  }
});
~~~



## Variations

* The hover effect can be hidden with `wash: false`.
* The ripple effect on click can be hidden with `ink: false`.
* Button contains no icon as this is not part of the Material Design guidelines; use [icon Button](../polythene-icon-button) instead



## Mobile and tap delay

To remove the tap delay on mobile devices it is advisable to use a library like [Fastclick](https://github.com/ftlabs/fastclick). But because Fastclick has an unresolved issue with tap events while scrolling on iOS, it is better to use the convenience wrapper provided in "polythene-fastclick". This temporarily removes the Fastclick event when an element is being scrolled.

~~~javascript
import "polythene-fastclick";
~~~



## Appearance

### Styling

Below are examples how to change the button appearance, either with a theme or with CSS.

You can find more information about theming in [Theme](../polythene-theme).

#### Themed component

~~~javascript
button.theme(".bordered-button", {
  color_light_text:   "#673ab7",
  color_light_border: "#673ab7",
  color_dark_text:    "yellow",
  color_dark_border:  "yellow"
});

m(button, {
  label: "Borders",
  className: "bordered-button",
  borders: true
});
~~~

#### CSS

Change CSS using the CSS Classes at the bottom of this page.

#### Style

Some style attributes can be set using option `style`. For example:

~~~javascript
m(button, {
  style: {
    backgroundColor: "#EF6C00",
    color:           "#fff"
  }
});
~~~

#### Dark or light tone

If the component - or a component's parent - has option `tone` set to "dark", the component will be rendered with light colors on dark. 

* Use `tone: "dark"` to render light on dark
* Use `tone: "light"` to locally render normally when dark tone is set



## Options

### Common component options

| **Parameter** |  **Mandatory** | **Type** | **Default** | **Description** |
| ------------- | -------------- | -------- | ----------- | --------------- |
| **element**   | optional       | String   | "a"         | HTML element tag; may also be "button" |
| **className** | optional       | String   |             | Extra CSS class appended to `pe-text-button` |
| **style**     | optional       | Object   |             | For setting simple style attributes |
| **id**        | optional       | String   |             | HTML element id |
| **content**   | use `label` or `content` | String, hyperscript or component | | Alternative content; replaces children and ignores `label` |
| **before**    | optional       | String, hyperscript or component | | Extra content before main content; note that this content is placed left of subsequent elements with a lower stacking depth |
| **after**     | optional       | String, hyperscript or component | | Extra content after main content; note that this content is placed right of preceding elements with a higher stacking depth |
| **events**    | optional       | Object | | Options object containing one or more standard events such as `onclick` |
| **tabindex**  | optional       | Integer | 0 | Tab index |
| **tone**      | optional       | String: "dark" or "light" |  | Renders the component light on dark (sets class `pe-dark-tone`); use "light" to locally inverse (sets class `pe-light-tone`) |

### Button specific options

| **Parameter** |  **Mandatory** | **Type** | **Default** | **Description** |
| ------------- | -------------- | -------- | ----------- | --------------- |
| **label** | use `label` or `content` | String | | The button label |
| **url** | optional | Object with `href`, optionally `oncreate` | | URL location; for in-app route linking set `oncreate : m.route.link` |
| **disabled** | optional | Boolean | false | Disables the button |
| **selected** | optional | Boolean | false | Set to true to show the button as selected |
| **formaction** | optional | String | | "The URI of a program that processes the information submitted by the button. If specified, it overrides the action attribute of the button"s form owner." [source:MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-formaction) |
| **inactive** | optional | Boolean | | Set to `true` to disable button events and ripple/wash effects |
| **inactivate** | optional | Number | | The number of seconds after tap/click when the button is inactive; useful to prevent double clicks |

### Button appearance options

| **Parameter** |  **Mandatory** | **Type** | **Default** | **Description** |
| ------------- | -------------- | -------- | ----------- | --------------- |
| **wash** | optional | Boolean | true | Set to false to hide the effect on hover |
| **ink** | optional | Boolean | true | Set to false to disable the ripple effect on click/tap |
| **ripple** | optional (valid if `ink` is `true`) | Options object | | Pass [ripple](../polythene-ripple) options to define ripple behavior |
| **increase** | optional | Number | 1 | The z-index increment/decrement on tap; note that the maximum z value is 5 |
| **borders** | optional | Boolean | false | Set to `true` to add a border; by default the border has no color - set border color to class `pe-button__content` to see the border |



### Composition

Button is composed from:

* [Ripple](../polythene-ripple) (when option `ink` is not `false`)



## CSS classes

See: `src/classes.js`



## Future

* Option to wait for ripple to finish before url/event is followed

