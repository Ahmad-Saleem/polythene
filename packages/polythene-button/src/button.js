import m from "mithril";
import { shadow } from "polythene-shadow";
import { ripple } from "polythene-ripple";
import { isTouch, touchStartEvent, touchEndEvent, subscribe, filterSupportedAttributes } from "polythene-core";
import { customTheme } from "./theme";

const classes = {
  component: "pe-button pe-button--text",
  content: "pe-button__content",
  label: "pe-button__label",
  raised: "pe-button--raised",
  wash: "pe-button__wash",
  focus: "pe-button__focus",
  selected: "pe-button--selected",
  disabled: "pe-button--disabled",
  borders: "pe-button--borders",
  inactive: "pe-button--inactive",
  focusState: "pe-button--focus"
};

const EL_ATTRS = [
  "formaction",
  "type"
];

const MAX_Z = 5;

let tapStart,
  tapEndAll = () => {},
  downButtons = [];

subscribe(touchEndEvent, () => tapEndAll());

const animateZ = (state, attrs, name) => {
  const zBase = state.zBase;
  const increase = attrs.increase || 1;
  let z = state.z;
  if (name === "down" && zBase !== 5) {
    z = Math.min(z + increase, MAX_Z);
  } else if (name === "up") {
    z = Math.max(z - increase, zBase);
  }
  if (z !== state.z) {
    state.z = z;
    m.redraw(); // show shadow animation
  }
};

const inactivate = (state, attrs) => {
  state.inactive = true;
  m.redraw();
  setTimeout(() => {
    state.inactive = false;
    m.redraw();
  }, attrs.inactivate * 1000);
};

const initTapEvents = (el, state, attrs) => {
  const tapHandler = (state, attrs, name) => {
    if (name === "down") {
      downButtons.push({
        state,
        attrs
      });
    } else
    if (name === "up") {
      if (attrs.inactivate && !state.inactive) {
        inactivate(state, attrs);
      }
    }
    // no z animation on touch
    const animateOnTap = attrs.animateOnTap !== false ? true : false;
    if (animateOnTap && !isTouch) {
      animateZ(state, attrs, name);
    }
  };
  tapStart = () => tapHandler(state, attrs, "down");
  tapEndAll = () => {
    downButtons.map((btn) => {
      tapHandler(btn.state, btn.attrs, "up");
    });
    downButtons = [];
  };
  el.addEventListener(touchStartEvent, tapStart);
};

const clearTapEvents = el => {
  el.removeEventListener(touchStartEvent, tapStart);
};

const view = vnode => {
  const state = vnode.state;
  const attrs = vnode.attrs;
  const noink = (attrs.ink !== undefined && attrs.ink === false);
  const disabled = attrs.disabled;
  const element = attrs.element || "a";
  const tabIndex = disabled || state.inactive
    ? -1
    : attrs.tabindex || 0;
  const onClickHandler = attrs.events && attrs.events.onclick;
  const props = Object.assign(
    {}, 
    filterSupportedAttributes(attrs, EL_ATTRS),
    {
      class: [
        (attrs.parentClass || classes.component),
        (attrs.selected ? classes.selected : null),
        (disabled ? classes.disabled : null),
        (state.inactive ? classes.inactive : null),
        (attrs.borders ? classes.borders : null),
        (attrs.raised ? classes.raised : null),
        (state.focus ? classes.focusState : null),
        attrs.class
      ].join(" "),
      tabIndex,
      // handle focus events
      onfocus: () => state.focus = !state.mouseover,
      onblur: () => state.focus = false,
      // don"t show focus on click (detected by not being in mouse over state)
      onmouseover: () => state.mouseover = true,
      onmouseout: () => state.mouseover = false,
      // if focus, dispatch click event on ENTER
      onkeydown: (e) => {
        if (e.which === 13 && state.focus) {
          state.focus = false;
          if (onClickHandler) {
            onClickHandler(e);
          }
        }
      }
    },
    attrs.events ? {...attrs.events} : null,
    attrs.url ? {...attrs.url} : null,
    disabled ? {disabled: true} : null
  );
  const label = attrs.content
    ? attrs.content
    : attrs.label
      ? typeof attrs.label === "object"
        ? attrs.label
        : m("div", {class: classes.label}, attrs.label)
      : vnode.children && vnode.children[0]
        ? vnode.children
        : null;
  const noWash = disabled || (attrs.wash !== undefined && !attrs.wash);
  const content = label
    ? m("div", {class: classes.content}, [
      attrs.raised && !disabled
        ? m(shadow, {z: state.z, animated: true})
        : null,
      // ripple
      disabled || noink
        ? null
        : m(ripple, attrs.ripple),
      // hover
      noWash ? null : m("div", {class: classes.wash}),
      // focus
      disabled ? null : m("div", {class: classes.focus}),
      label
    ])
    : null;
  return m(element, props, [attrs.before, content, attrs.after]);
};

export const button = {
  theme: customTheme, // accepts (className, vars)
  oninit: vnode => {
    const z = (vnode.attrs.z !== undefined) ? vnode.attrs.z : 1;
    vnode.state = {
      el: undefined,
      zBase: z,
      z: z,
      tapEventsInited: false,
      inactive: !!vnode.attrs.inactive,
      focus: false,
      mouseover: false
    };
  },
  oncreate: vnode => {
    if (!vnode.attrs.disabled && !vnode.state.inactive && !vnode.state.tapEventsInited) {
      vnode.state.el = vnode.dom;
      initTapEvents(vnode.dom, vnode.state, vnode.attrs);
      vnode.state.tapEventsInited = true;
    }
  },
  onremove: vnode => {
    if (vnode.state.tapEventsInited) {
      clearTapEvents(vnode.dom);
    }
  },
  view
};
