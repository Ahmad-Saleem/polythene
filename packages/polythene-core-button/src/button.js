import { filterSupportedAttributes } from "polythene-core-essentials";
import { customTheme } from "./theme";
import classes from "./classes";

export const element = "a";

export const theme = customTheme;

export const createProps = (vnode, { updateState, keys: k }) => {
  const state = vnode.state;
  const attrs = vnode.attrs;
  const disabled = attrs.disabled;
  const inactive = attrs.inactive || state.inactive;
  const onClickHandler = attrs.events && attrs.events.onclick;
  const handleInactivate = () => (
    updateState("inactive", true),
    setTimeout(() => (
      updateState("inactive", false)
    ), attrs.inactivate * 1000)
  );
  return Object.assign(
    {}, 
    filterSupportedAttributes(attrs, { add: [k.formaction, "type"] }),
    {
      className: [
        attrs.parentClass || classes.component,
        attrs.selected ? classes.selected : null,
        disabled ? classes.disabled : null,
        inactive ? classes.inactive : null,
        attrs.borders ? classes.borders : null,
        state.focus ? classes.focused : null,
        attrs.tone === "dark" ? "pe-dark-tone" : null,
        attrs.tone === "light" ? "pe-light-tone" : null,
        attrs.className || attrs[k.class],
      ].join(" ")
    },
    inactive ? null : {
      [k.tabindex]: disabled || inactive
        ? -1
        : attrs[k.tabindex] || 0,
      [k.onclick]: e => (
        attrs.inactivate !== undefined && handleInactivate(),
        onClickHandler && onClickHandler(e),
        true
      ),
      // handle focus events
      [k.onfocus]: () => updateState("focus", !state.mouseover),
      [k.onblur]: () => updateState("focus", false),
      // don't show focus on click (detected by not being in mouse over state)
      [k.onmouseover]: () => updateState("mouseover", true),
      [k.onmouseout]: () => updateState("mouseover", false),
      // if focus, dispatch click event on ENTER
      [k.onkeydown]: e => {
        if (e.which === 13 && state.focus) {
          updateState("focus", false);
          if (onClickHandler) {
            onClickHandler(e);
          }
        }
      }
    },
    attrs.style ? { style: {} } : null, // Set style on content, not on component
    attrs.events,
    attrs.url,
    disabled ? { disabled: true } : null
  );
};

export const createContent = (vnode, { renderer: h, keys: k, ripple }) => {
  const attrs = vnode.attrs;
  const noink = attrs.ink !== undefined && attrs.ink === false;
  const disabled = attrs.disabled;
  const children = attrs.children || vnode.children;
  const label = attrs.content
    ? attrs.content
    : attrs.label
      ? typeof attrs.label === "object"
        ? attrs.label
        : h("div", { key: "label", className: classes.label }, attrs.label)
      : children
        ? children
        : null;
  const noWash = disabled || (attrs.wash !== undefined && !attrs.wash);
  return label
    ? h("div", {
      [k.class]: classes.content,
      key: "button",
      style: attrs.style || {}
    }, [
      !disabled && attrs.shadowComponent // "protected" option, used by raised-button
        ? attrs.shadowComponent
        : null,
      // ripple
      disabled || noink
        ? null
        : ripple && h(ripple, {
          ...attrs.ripple,
          key: "ripple",
          // target: () => vnode.dom
        }),
      // hover
      noWash ? null : h("div", { key: "wash", className: classes.wash }),
      // focus
      disabled ? null : h("div", { key: "focus", className: classes.focus }),
      label
    ])
    : null;
};
