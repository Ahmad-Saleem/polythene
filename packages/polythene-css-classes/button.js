
export default {
  base:       "pe-button",
  component:  "pe-button pe-text-button",
  row:        "pe-button-row",

  // elements
  content:    "pe-button__content",
  focus:      "pe-button__focus",
  label:      "pe-button__label",
  wash:       "pe-button__wash",

  // states
  borders:    "pe-button--borders",
  disabled:   "pe-button--disabled",
  focused:    "pe-button--focus",
  inactive:   "pe-button--inactive",
  selected:   "pe-button--selected",
};
