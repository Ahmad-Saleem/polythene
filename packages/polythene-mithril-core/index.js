
export { keyer, renderer } from "./src";
