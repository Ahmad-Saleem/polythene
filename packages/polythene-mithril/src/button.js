import { statefulComponent } from "./core";
import { button as component } from "polythene-new-core";
import { ripple } from "./ripple";

const createProps = (vnode, args) => component.createProps(vnode, Object.assign(args, { ripple }));
const createContent = (vnode, args) => component.createContent(vnode, Object.assign(args, { ripple }));

export const button = statefulComponent(Object.assign(
  {},
  component,
  {
    createProps,
    createContent
  }
));

button.theme = component.theme;
