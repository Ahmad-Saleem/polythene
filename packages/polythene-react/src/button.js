import { stateComponent } from "./core";
import { button as component } from "polythene-new-core";
import { ripple } from "./ripple";

const createProps = (vnode, args) => component.createProps(vnode, Object.assign(args, {ripple}));
const createContent = (vnode, args) => component.createContent(vnode, Object.assign(args, {ripple}));

export const button = stateComponent({
  createProps,
  createContent,
  element: component.element
});

button.theme = component.theme;
