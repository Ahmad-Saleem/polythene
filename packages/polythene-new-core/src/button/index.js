import { createProps, createContent, theme, element } from "./button";
import classes from "./classes";
import vars from "./theme/vars";

export const button = {
  createProps, createContent, theme, element,
  classes,
  vars
};
