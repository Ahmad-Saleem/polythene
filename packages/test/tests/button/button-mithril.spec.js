import { runSnapshots } from "../../scripts/mithril-snapshots";
import { renderer } from "polythene-mithril-core";
import button from "polythene-mithril-button";
import tests from "./spec-tests.js";

runSnapshots({
  tests: tests({ button, renderer }),
  renderer
});
