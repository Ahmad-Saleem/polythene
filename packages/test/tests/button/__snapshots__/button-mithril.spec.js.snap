// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No options 1`] = `
"<a class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;

exports[`Option: after 1`] = `
"<a class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div><span>After</span></a>
"
`;

exports[`Option: before 1`] = `
"<a class=\\"pe-button pe-text-button\\" tabindex=
\\"0\\"><span>Before</span>
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;

exports[`Option: children 1`] = `
"<a class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div><span>Child</span>
</div></a>
"
`;

exports[`Option: class 1`] = `
"<a class=\\"pe-button pe-text-button class-x\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;

exports[`Option: className 1`] = `
"<a class=\\"pe-button pe-text-button className-x\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;

exports[`Option: content 1`] = `
"<a class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div><span>Content</span>
</div></a>
"
`;

exports[`Option: element 1`] = `
"<button class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></button>
"
`;

exports[`Option: id 1`] = `
"<a id=\\"id-x\\" class=\\"pe-button pe-text-button\\" tabindex=\\"0\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;

exports[`Option: tabindex 1`] = `
"<a tabindex=\\"3\\" class=\\"pe-button pe-text-button\\">
<div class=\\"pe-button__content\\">
  <div class=\\"pe-ripple pe-ripple--constrained\\">
    <div class=\\"pe-ripple__mask\\">
      <div class=\\"pe-ripple__waves\\"></div>
    </div>
  </div>
  <div class=\\"pe-button__wash\\"></div>
  <div class=\\"pe-button__focus\\"></div>
</div></a>
"
`;
